skills/src/lasr_skills/vision/get_cropped_image.py

```diff
@@ -0,0 +1,227 @@
+
+        detection = None
+        dist = float("inf")
+
+        for det in detections:
+            x, y, _, _ = det.xywh[0], det.xywh[1], det.xywh[2], det.xywh[3]
+            det_center_x = x
+            det_center_y = y
+            det_dist = np.sqrt(
+                (x_to_compare - det_center_x) ** 2 + (y_to_compare - det_center_y) ** 2
+            )
+            if det_dist < dist:
+                dist = det_dist
+                detection = det
+
+        if detection is None:
+            raise ValueError("No detections found")
+
+        x, y, w, h = (
+            detection.xywh[0],
+            detection.xywh[1],
+            detection.xywh[2],
+            detection.xywh[3],
+        )
+        return image[y - h // 2 : y + h // 2, x - w // 2 : x + w // 2]
+
+    def _3d_crop(
+        self, pointcloud: np.ndarray, robot_loc: Point, detections: List[Detection3D]
+    ) -> np.ndarray:
+        """Crops the image to the desired object that is closest to the
+        centroid of the image.
+
+        Args:
+            pointcloud (np.ndarray): pointcloud from depth camera; RGB will be extracted
+            from this.
+            robot_loc (Point): Current location of the robot, used to calculate
+            3D-distance.
+            detections (YoloDetection): Yolo Detections of the desired object
+            in the image.
+        Returns:
+            np.ndarray: Cropped image
+        """
+        min_dist = float("inf")
+        max_dist = 0
+        closest_detection = None
+        furthest_detection = None
+        for det in detections:
+            dist = np.sqrt(
+                (robot_loc.x - det.point.x) ** 2
+                + (robot_loc.y - det.point.y) ** 2
+                + (robot_loc.z - det.point.z) ** 2
+            )
+            if dist < min_dist:
+                min_dist = dist
+                closest_detection = det
+            if dist > max_dist:
+                max_dist = dist
+                furthest_detection = det
+
+        if self._crop_method == "closest":
+            detection = closest_detection
+        elif self._crop_method == "furthest":
+            detection = furthest_detection
+        else:
+            raise ValueError(f"Invalid 3D crop_method: {self._crop_method}")
+
+        if detection is None:
+            raise ValueError(f"No detection found")
+
+        rgb_image = pcl_to_cv2(pointcloud)
+        x, y, w, h = (
+            detection.xywh[0],
+            detection.xywh[1],
+            detection.xywh[2],
+            detection.xywh[3],
+        )
+
+        return rgb_image[y - h // 2 : y + h // 2, x - w // 2 : x + w // 2]
+
+    def execute(self, userdata):
+        try:
+            if self._crop_method in self._valid_2d_methods:
+                img_msg = rospy.wait_for_message(self._rgb_topic, Image)
+                detections = self._yolo_2d(
+                    img_msg, "yolov8x.pt", 0.5, 0.3
```

Comment:
Maybe the model, confidence and nms threshold should be given in the constructor?